mysite/templates/ct/lessons.html

```diff
@@ -136,28 +136,110 @@
 </div>
 {% endif %}
 
-{% if lessonSet %}
+{#{% if lessonSet %}#}
+{#<h2>Search Results</h2>#}
+{#If one of these is relevant to this concept,#}
+{#please click <b>{{ actionLabel }}</b>.#}
+{#<table class="table table-striped">#}
+{#<thead><tr>#}
+{#  <th>Lesson</th><th>Type</th><th>Author</th>#}
+{#</tr></thead>#}
+{#<tbody>#}
+{#{% for ul in lessonSet %}#}
+{#  <tr><td>#}
+{#  <a href="{{ actionTarget |get_object_url:ul }}">{{ ul.lesson.title }} {{ ul.treeID }} {{ ul.branch }}</a>#}
+{#  <form action="{{ actionTarget }}" method="post"#}
+{#   style=" display:inline!important;">#}
+{#  {% csrf_token %}#}
+{#  <input type="hidden" name="ulID" value="{{ ul.pk }}" />#}
+{#  <input type="submit" value="{{ actionLabel }}" />#}
+{#  </form>#}
+{#  </td>#}
+{#    <td>{{ ul.lesson.get_kind_display }}</td>#}
+{#    <td>{{ ul.lesson.addedBy.get_full_name }}</td>#}
+{#  </tr>#}
+{#{% endfor %}#}
+{#</tbody>#}
+{#</table>#}
+{#{% elif foundNothing %}#}
+{#  <b>No matches found</b>.  Try another search, or write a new lesson.#}
+{#{% endif %}#}
+
+{% if full_list %}
 <h2>Search Results</h2>
 If one of these is relevant to this concept,
 please click <b>{{ actionLabel }}</b>.
 <table class="table table-striped">
 <thead><tr>
-  <th>Lesson</th><th>Type</th><th>Author</th>
+  <th>Lesson</th><th></th><th>Type</th><th>Author</th>
 </tr></thead>
 <tbody>
-{% for ul in lessonSet %}
-  <tr><td>
-  <a href="{{ actionTarget |get_object_url:ul }}">{{ ul.lesson.title }}</a>
-  <form action="{{ actionTarget }}" method="post"
-   style=" display:inline!important;">
-  {% csrf_token %}
-  <input type="hidden" name="ulID" value="{{ ul.pk }}" />
-  <input type="submit" value="{{ actionLabel }}" />
-  </form>
-  </td>
-    <td>{{ ul.lesson.get_kind_display }}</td>
-    <td>{{ ul.lesson.addedBy.get_full_name }}</td>
-  </tr>
+{% for tree, branch in full_list.items %}
+    {% if branch %}
+        <tr><td colspan="4">
```

Comment:
Just an idea. Extract this section into separate sub-template and `{% include %}` it here.

Because right now this if-else statement is almost too long.